ui: add locality navigation to clusterviz

couchand opened the pull request:
Makes the localities navigable.  Clearly still lots more to do here, but baby steps.

couchand committed:
ui: add locality navigation to clusterviz

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22303)
<!-- Reviewable:end -->

vilterp commented:
No back button in the UI yet, but the browser back button works! Ayyyyy

vilterp commented on pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx:
```diff
@@ -6,106 +6,143 @@
+interface ModalLocalitiesViewProps {
+  localityTree: LocalityTree;
+  locationTree: LocationTree;
+  tiers: LocalityTier[];
```
naming quibble: would prefer to call this localityPath or something, because it's the path we've walked down in the locality tree.

I understand that you're trying to match how nodes have "tiers" on them though.

vilterp reviewed (approved):
Looks good, just a naming quibble. Feel free to ignore, merge & move on

mrtracy commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 34 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22303#-L4HmvhA-_fEM9Kihzpr:-L4HmvhA-_fEM9Kihzps:b-ktb1lm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L34)):*
> ```XML
> 
>   onClick = () => {
>     const tiers = this.props.locality.tiers.map(({ key, value }) => key + "=" + value).join("/");
> ```

You've used this pattern at least twice, `{key, value}[] -> string`; might be a good method to add right next to "parseLocalityRoute".

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 81 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22303#-L4HoLmWDfPlqBsV8QCP:-L4HoLmXDYJeoVrjxnfH:b-jake00) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L81)):*
> ```XML
> class NodeBox extends React.Component<NodeBoxProps, any> {
>   render() {
>     const tiers = this.props.node.desc.locality.tiers.map(({ key, value }) => ({ key, value }));
> ```

What is this line doing?  Just removing other properties?

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 83 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22303#-L4HoWgu5A4xle3rqWCI:-L4HoWgu5A4xle3rqWCJ:bkpeakc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L83)):*
> ```XML
>     const tiers = this.props.node.desc.locality.tiers.map(({ key, value }) => ({ key, value }));
>     const location = findMostSpecificLocation(this.props.locationTree, tiers);
>     const center = this.props.projection([location.longitude, location.latitude]);
> ```

This subroutine (find center given locationTree and tiers) is repeated in both NodeBox and LocalityBox.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 110 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22303#-L4HpikL1qYjqXTW9Ftl:-L4HpikL1qYjqXTW9Ftm:b-diz6uo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L110)):*
> ```XML
> 
>     Object.keys(tree.localities).forEach((key) => {
>       Object.keys(tree.localities[key]).forEach((value) => {
> ```

Unless i'm missing something, I think this would be more readable as ```
```typescript
_.values(tree.localities).each(locality => {
   _.values(locality).each(child => {
     children.push(...);
   })
});
```

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 139 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22303#-L4Hr0Xm98xS2O7aLi6a:-L4Hr0Xm98xS2O7aLi6b:bmtjxj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L139)):*
> ```XML
>     let treeToRender = this.props.localityTree;
>     this.props.tiers.forEach(({ key, value }) => {
>       treeToRender = treeToRender.localities[key][value];
> ```

Looks like you could easily get a null reference exception here if you typed in a bad URL.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22303)*
<!-- Sent from Reviewable.io -->

couchand committed:
ui: refactor some locality handling

Release note: None

couchand committed:
ui: refactor renderChildLocalities

Release note: None

couchand committed:
ui: move vector to util folder

Release note: None

couchand committed:
ui: invert location finding for clusterviz

The location to plot a locality at is determined by the localities under it,
not the localities above it.  Nodes should never be plotted on the map.

Release note: None

couchand commented:
Review status: 0 of 12 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303#-L4HmvhA-_fEM9Kihzpr:-L4I9oMyBNXC-4Z-odqd:bkrzmdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L34)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

You've used this pattern at least twice, `{key, value}[] -> string`; might be a good method to add right next to "parseLocalityRoute".
</blockquote></details>

good idea, particularly since it is doing exactly the opposite.  done.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303#-L4HoLmWDfPlqBsV8QCP:-L4I9sly1bGGq1apgi9G:bh0lh2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L81)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

What is this line doing?  Just removing other properties?
</blockquote></details>

clearing up the discrepancy around the requiredness of these props.  protobug.js marks every prop as optional, which is really annoying.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 83 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303#-L4HoWgu5A4xle3rqWCI:-L4IA0vHUx-2zEncaYbm:bvy0cl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L83)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This subroutine (find center given locationTree and tiers) is repeated in both NodeBox and LocalityBox.
</blockquote></details>

yep.  it's not actually relevant to nodes, since they shouldn't ever be mapped, since the locations are assigned to localities, not nodes.  i've removed it from there, but still haven't implemented the right layout for them.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303#-L4HfOok_GTe6mJ8Ezor-r1-99:-L4IAAMRmue99v1j6YbL:b-mx77jz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L99)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

naming quibble: would prefer to call this localityPath or something, because it's the path we've walked down in the locality tree.

I understand that you're trying to match how nodes have "tiers" on them though.
</blockquote></details>

noted.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303#-L4HpikL1qYjqXTW9Ftl:-L4IAMYOj9lBiWE6N8df:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L110)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Unless i'm missing something, I think this would be more readable as ```
```typescript
_.values(tree.localities).each(locality => {
   _.values(locality).each(child => {
     children.push(...);
   })
});
```
</blockquote></details>

done.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303#-L4Hr0Xm98xS2O7aLi6a:-L4IANb5JA_ioojHmcj6:bbnzt7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L139)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Looks like you could easily get a null reference exception here if you typed in a bad URL.
</blockquote></details>

yep.  ignoring for now, need to handle that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 12 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx, line 139 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303#-L4Hr0Xm98xS2O7aLi6a:-L4IEyqeH_b2YyNTbwfy:bn245x0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5e1d67202015747c78d8770c52eae0ca2c37b79/pkg/ui/ccl/src/views/clusterviz/containers/map/modalLocalities.tsx#L139)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

yep.  ignoring for now, need to handle that.
</blockquote></details>

If you do another pass to squash, maybe add a TODO. Otherwise, I can trust you to test this out when you get to a later stage.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303#-:-L4IF95OfbU9iZIg5efc:bnfp4nl)*
<!-- Sent from Reviewable.io -->

couchand committed:
ui: findOrCalculateLocation should never look up

Release note: None

vilterp commented on pkg/ui/src/util/locations.spec.ts:
```diff
@@ -60,3 +60,125 @@ describe("findMostSpecificLocation", function() {
+              desc: {
+                node_id: 1,
+                locality: {
+                  tiers: nycLocality
```
^ missing trailing comma here is what's failing the build, lol

couchand committed:
ui: redirect to clusterviz on locality route errors

Release note: None

couchand commented:
@mrtracy i went ahead and just fixed the routing bug so i wouldn't forget.
@vilterp thanks for the heads up, fixed.

PTAL

mrtracy commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for fixing that, I'll rest much easier

---

Review status: 0 of 13 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22303#-:-L4b7Gy_ymEQu4D2BUeF:b-jsc2vb)*
<!-- Sent from Reviewable.io -->
